### ISSUE TITLE
Qdrant on ECS, with efs storage mounted: Service internal error: RocksDB open error: IO error: While lock file, Resource temporarily unavailable

### ISSUE THREAD

**Issue by msciancalepore98** [qdrant/qdrant#4145]
I have deployed a Qdrant instance on ECS, where the storage path is mounted on an EFS disk. 
Scenario: The first deploy goes just fine, everything works; If for some reason, I re-trigger the ECS task deployment, a RocksDB-related problem occurs:

```
Panic occurred in file /qdrant/lib/collection/src/shards/replica_set/mod.rs at line 261: Failed to load local shard "./storage/collections/test-collection/0": Service internal error: RocksDB open error: IO error: While lock file: ./storage/collections/test-collection/0/segments/6ccb0e5a-6176-49d0-8eb7-6c5eb4b0d3b8/LOCK: Resource temporarily unavailable
```
From this error, it seems that a rolling update of Qdrant on ECS/k8s could never work, due to the fact that the first replica is attached to the storage, and concurrently the new replica tries to attach to the same storage in R/W.. am I missing anything here? Also, it seems [that](https://github.com/facebook/rocksdb/issues/991) only one RocksDB process can access the same DB... Right now this is a big blocker and I didn't find any solutions to this, any ideas?

(I mean, a basic solution would be to store data in the container virtual storage, but I would lose all the vectors if the container restarts..)

**Comment by generall** [comment]
How do you deploy qdrant in your k8s?

We have several ready-made solutions like https://github.com/qdrant/qdrant-helm or hybrid-cloud https://hybrid-cloud.qdrant.tech/ where those problems are all resolved

**Comment by msciancalepore98** [comment]
It's just a simple ECS task deploy, where EBS is used to provide persistency (efs). Is it possible that I need a non sharable disk? I saw in that helm example that the PVC is of type ReadWriteOnce..

**Comment by generall** [comment]
Hey @msciancalepore98, I can't give you any guarantees of qdrant's work and what is expected to happen or not, if you continue to butchering storage internals like this.

**Comment by msciancalepore98** [comment]
If you could actually help with proper debugging hints would be great as well, I am trying different things to get to the root cause of this behaviour with EFS.

Also, I can't use any auto-managed solution in my environment, only deploying tasks on ECS.

**Comment by ryanlee588** [comment]
I am facing a similar issue. Commenting to stay to date

**Comment by timvisee** [comment]
> Is it possible that I need a non sharable disk?

Correct. At least, in terms of file shares, we do recommend not to use this.

Also, each instance must have their very own storage directory. These cannot be shared. The cluster itself will take care of putting sharing all your data across the cluster and putting it in each storage directory separately.

> We never tested qdrant on EFS

@generall In their [FAQ](https://aws.amazon.com/efs/faq/) they do promise strong consistency and support for proper file locking. But I also feel like we've seen issues with this before.

**Comment by timvisee** [comment]
And this happens on every restart, and you're 100% sure you don't have another instance running on the same data?

To be honest, I'm not entirely sure. We haven't hit this ourselves yet.

You might end up having to purge lock files yourself, but I have no idea what other damage that might do.

**Comment by pvieito** [comment]
Hey @timvisee @generall:

> And this happens on every restart, and you're 100% sure you don't have another instance running on the same data?

This is an issue for example when you deploy Qdrant in a service that automatically monitors & relaunches it on failure, like ECS or Kubernetes. For example, imagine that Kubernetes is doing a health-check on the Qdrant endpoint, it starts to fail and it launches a new Qdrant to replace the old one, it connects it to the same storage but it has the LOCKs from the failed instance. Qdrant should have some sort of env-var or configuration to do a clean-up on start and remove any locks from previous failed instances / runs.

**Comment by timvisee** [comment]
> Qdrant should have some sort of env-var or configuration to do a clean-up on start and remove any locks from previous failed instances / runs.

As far as I'm aware, it does this already.

Running locally and killing with `kill -9` doesn't show this. We don't see this problem in normal k8s operation either. That's why I wonder whether locking on EFS is as good as the promise it to be.

I'll try to do some debugging later to see whether I can catch the same problem.

**Comment by sutsr** [comment]
Has there been any progress on determining what is going awry here?

I observe the same behaviour with a single node cluster operating on a single dedicated EFS store: namely one of old LOCK files not being cleared when a new deployment is run in ECS.

It makes sense that the LOCK file warning persists during an ECS redeployment, as AWS tries to bring up a second copy of the task to green status before terminating the old version. During termination of the old version I assume intended behaviour is that the associated LOCK files (which were under the old instance's control, not the new instance) are cleared?

To add an extra detail, I also intermittently observe that this warning and associated panic cycling resolves itself many hours later, and the cluster is actually up and responsive most of the time during that cycling process (it occurring approx. every 11 mins).

**Comment by msciancalepore98** [comment]
with the current status, there is no fix this with blue-green deployments, like what ECS does.
My only feasible "fix" is to avoid redeploying a new instance. yeah..
If a redeploy happens, be sure to hard delete the ECS task and wait for its termination, which takes tens of minutes.
it just sucks.
